### PR TITLE
Dynamic tile arguments

### LINKED ISSUE #240: Cumulative view of user tracks
This issue is one of the feature requests from [this post](https://forum.openbikesensor.org/t/ich-wuensch-mir-ein-paar-neue-geaenderte-features/841):

Klaus writes:  Es gibt im Portal einen Knopf mit dem ich mir „meine“ Tracks anzeigen lassen kann. Ich fände auch einen Knopf mit denen ich mir „meine“ Messdaten auf der Karte anschauen könnte, hilfreich.

I spoke to him and the "nicest" rendering would of course be a map showing exclusively the events from the logged-in users own tracks.

I looked into the database and we have the data we would need for this filtering in the portal.

A possible side advantage would be comparing different drivers on the same roads - i.e. Klaus on his road bike vs me on my recumbent vs Anna on the cargo bike - But cooperation of the users would be required as this is obviously personal data, so only ones own data should be available.

### PR DESCRIPTION
**Pull request by opatut**
Adds filters to the map to dynamically limit the data being parsed.

Fixes #240, #241.

### CODE REVIEW

File: api/obs/api/routes/tiles.py [hunk 1]
```diff
@@ -38,11 +41,27 @@ async def tiles(req, zoom: int, x: int, y: str):
         tile = get_tile(req.app.config.TILES_FILE, int(zoom), int(x), int(y))
 
     else:
+        user_id = None
+        username = req.ctx.get_single_arg("user", default=None)
+        if username is not None:
+            if req.ctx.user is None or req.ctx.user.username != username:
+                raise Forbidden()
```
Comment (gluap) [inline api/obs/api/routes/tiles.py:48]: Tested this before even looking at the code :) Was not disappointed to get a "forbidden".

**Review by gluap — CHANGES_REQUESTED** [review]
Tested this on my machine, including trying to fetch the private tiles for another user (and being presented with a forbidden error, :no_entry_sign: :+1: as it should be).

There's one privacy issue though: Allowing a time range as narrow as one day allows to see (anonymous) track data. Scenario: see someone riding with an OBS, later narrow the interval to just the day they were riding, switch to "road usage view" and see their (mostly continuous) track. Unless of course there are tons of riders in the area on that day - but that's unlikely on most portal instances.  Example of such a track track that (from the beginning and end not shown in the picture) I was able to identify as mine.

image

We could combat this by making the selector less precise - Maybe only allowing to select a month, but no day. 

The same with more inbetween steps will be possible with the "before-after" mode by subtracting before and after from the total - also arriving at the usage numbers in the interval between before and after.

I love the functionality though and think it is pretty powerful as a tool for investigation of the impacts of construction works and other temporal changes on the overtaking distance.

**Comment by opatut** [comment]
Yeah I am unsure about that fact. However, the same is already possible by showing events on the map, which come with the timestamp, so you can reconstruct movement patterns from them, provided there is little data overlap with other tracks (at the moment quite common). Sure, this makes it easier to visualize that, but the data is already there.

In your attack scenario, you still need the mapping of the known timestamp to the user you've seen in real life to be able to extract information, so the leak (mapping a track to a human person) is technically happening outside the platform. Doesn't make it less valid, but harder to combat...

I'd say we round the date to weeks for now (as construction work will often align to weeks), what do you say?

**Review by gluap — APPROVED** [review]
Agreed - Rounding to weeks allows for decent anonymization with decent granularity in evaluation.

**Pull request merged**